ports/nrf/common-hal/busio/I2C.c

```diff
@@ -0,0 +1,207 @@
+                                    | ((uint32_t)GPIO_PIN_CNF_SENSE_Disabled   << GPIO_PIN_CNF_SENSE_Pos);
+
+    NRF_GPIO->PIN_CNF[sda->pin] = ((uint32_t)GPIO_PIN_CNF_DIR_Input        << GPIO_PIN_CNF_DIR_Pos)
+                                    | ((uint32_t)GPIO_PIN_CNF_INPUT_Disconnect << GPIO_PIN_CNF_INPUT_Pos)
+                                    | ((uint32_t)GPIO_PIN_CNF_PULL_Disabled    << GPIO_PIN_CNF_PULL_Pos)
+                                    | ((uint32_t)GPIO_PIN_CNF_DRIVE_S0S1       << GPIO_PIN_CNF_DRIVE_Pos)
+                                    | ((uint32_t)GPIO_PIN_CNF_SENSE_Disabled   << GPIO_PIN_CNF_SENSE_Pos);
+
+    // 1 for I2C, 0 for SPI
+    self->twi = NRF_TWIM1;
+
+    if ( frequency < 100000 ) {
+      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K100;
+    }else if ( frequency < 250000 ) {
+      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K250;
+    }else {
+      self->twi->FREQUENCY = TWIM_FREQUENCY_FREQUENCY_K400;
+    }
+
+    self->twi->ENABLE = (TWIM_ENABLE_ENABLE_Enabled << TWIM_ENABLE_ENABLE_Pos);
+
+    self->twi->PSEL.SCL = scl->pin;
+    self->twi->PSEL.SDA = sda->pin;
+
+}
+
+bool common_hal_busio_i2c_deinited(busio_i2c_obj_t *self) {
+  return self->twi->ENABLE == 0;
+}
+
+void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
+    if (common_hal_busio_i2c_deinited(self)) {
+        return;
+    }
+
+    uint8_t scl_pin = self->twi->PSEL.SCL;
+    uint8_t sda_pin = self->twi->PSEL.SDA;
+
+    self->twi->ENABLE   = (TWIM_ENABLE_ENABLE_Disabled << TWIM_ENABLE_ENABLE_Pos);
+    self->twi->PSEL.SCL = (TWIM_PSEL_SCL_CONNECT_Disconnected << TWIM_PSEL_SCL_CONNECT_Pos);
+    self->twi->PSEL.SDA = (TWIM_PSEL_SDA_CONNECT_Disconnected << TWIM_PSEL_SDA_CONNECT_Pos);
+
+    reset_pin(scl_pin);
+    reset_pin(sda_pin);
+}
+
+bool common_hal_busio_i2c_probe(busio_i2c_obj_t *self, uint8_t addr) {
+  // Write no data when just probing
+  return 0 == common_hal_busio_i2c_write(self, addr, NULL, 0, true);
+}
+
+bool common_hal_busio_i2c_try_lock(busio_i2c_obj_t *self) {
+    bool grabbed_lock = false;
+//    CRITICAL_SECTION_ENTER()
+        if (!self->has_lock) {
+            grabbed_lock = true;
+            self->has_lock = true;
+        }
+//    CRITICAL_SECTION_LEAVE();
+    return grabbed_lock;
+}
+
+bool common_hal_busio_i2c_has_lock(busio_i2c_obj_t *self) {
+    return self->has_lock;
+}
+
+void common_hal_busio_i2c_unlock(busio_i2c_obj_t *self) {
+    self->has_lock = false;
+}
+
+uint8_t common_hal_busio_i2c_write(busio_i2c_obj_t *self, uint16_t addr, const uint8_t *data, size_t len, bool stopBit) {
+  NRF_TWIM_Type* twi = self->twi;
+
+  twi->ADDRESS       = addr;
+  twi->TASKS_RESUME  = 1;
+
+  twi->TXD.PTR       = (uint32_t) data;
+  twi->TXD.MAXCNT    = len;
+
+  twi->TASKS_STARTTX = 1;
+
+  // Wait for TX started
+  while(!twi->EVENTS_TXSTARTED && !twi->EVENTS_ERROR) {}
+  twi->EVENTS_TXSTARTED = 0;
+
+  // Wait for TX complete
+  if ( len )
+  {
```

Comment:
FYI my style is usually `if (len) {`. I won't nitpick it though until we can auto-validate it.